### PR TITLE
Check file size of `snapshot_version` when unarchiving snapshot

### LINKED ISSUE #18841: Check file size of `snapshot_version` when unarchiving snapshot
Related to #7167

_Original comment from original PR:_

> also, I just noticed we're better off checking file size of `snapshot_version`? for eventual untrusted snapshot thing?
> 
> Otherwise, it could be dos vector if crafted snapshot contains a 100G zeroed snapshot_version file because we read all to memory first....
> 
> unrelated to this pr off course. so could be done as a follow up pr...

__Originally posted by @ryoqun in https://github.com/solana-labs/solana/pull/18504#discussion_r674932737__

### PR DESCRIPTION
**Pull request by mooori**
#### Problem

The size of the `snapshot_version` file is not checked, which is an attack vector. See #18504, #18841. 

#### Summary of Changes

Check the size of the file containing the `snapshot_version` before opening it. If the size exceeds `MAX_SNAPSHOT_VERSION_FILE_SIZE`, an error is returned and the file will not be opened.

#### Notes

- `MAX_SNAPSHOT_VERSION_FILE_SIZE` was chosen arbitrarily - let me know if another value makes more sense.
- If the file size is too big, an error is created and returned according to other code in `snapshot_utils` dealing with files that are too big (namely `create_snapshot_data_file_stream`).

#### Test

So far, there is no dedicated test for `unarchive_snapshot`. Setting up such a test just to verify the `snapshot_version` file size is checked correctly looks like overkill.

Fixes #18841

### CODE REVIEW

**Comment by mvines** [comment]
Oh nice. This PR lgtm but @brooksprumo can you please take a look as well for a full review

**Comment by brooksprumo** [comment]
@mooori Thank you for your submission! Please mark the PR as "Ready for review" when you feel it's completed and I will take a look.

**Comment by mooori** [comment]
@brooksprumo CI checks haven't finished yet, but since it's just a small PR I think it's good to go (status is updated).

If there should be any issues, I'll be back later. I ran `cargo test`, `cargo fmt` and `cargo clippy` locally inside `runtime` and they didn't complain.

File: runtime/src/snapshot_utils.rs [hunk 1]
```diff
@@ -965,6 +966,19 @@ where
...
+            let error_message = format!(
+                "snapshot version file too large: {} has {} bytes (max size is {} bytes)",
+                unpacked_version_file.display(),
+                file_size,
+                MAX_SNAPSHOT_VERSION_FILE_SIZE,
+            );
+            return Err(get_io_error(&error_message));
+        }
+
+        // Read snapshot_version from file.
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:981]: Please move this into its own function (and all the way down to line 986, the end of this block, but I cannot select that far in GitHub...).

Named something like `snapshot_version_from_file()` and have it take an `impl AsRef<Path>` and return `Result<String>`.

File: runtime/src/snapshot_utils.rs [hunk 2]
```diff
@@ -48,6 +48,7 @@ pub const SNAPSHOT_STATUS_CACHE_FILE_NAME: &str = "status_cache";
 pub const DEFAULT_FULL_SNAPSHOT_ARCHIVE_INTERVAL_SLOTS: Slot = 100_000;
 pub const DEFAULT_INCREMENTAL_SNAPSHOT_ARCHIVE_INTERVAL_SLOTS: Slot = 100;
 const MAX_SNAPSHOT_DATA_FILE_SIZE: u64 = 32 * 1024 * 1024 * 1024; // 32 GiB
+const MAX_SNAPSHOT_VERSION_FILE_SIZE: u64 = 128; // byte
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:51]: I'd prefer this constant to be smaller. Since the version is only 5 bytes now, I propose this value be set to 8.

**Review by brooksprumo — CHANGES_REQUESTED** [review]
Overall looks good! I'm requesting some small tweaks.

Also:
> So far, there is no dedicated test for unarchive_snapshot. Setting up such a test just to verify the snapshot_version file size is checked correctly looks like overkill.

I'd like tests for the file-to-snapshot-version function (see comments below). Those tests do not necessarily need to be part of this PR though. If you'd rather not add the tests here, please create a new issue name something like "Add tests for snapshot_version_from_file" and mark it as "good first issue".

I'd like two tests:

1. Ensure that a version file with the default `SnapshotVersion` passes
2. Ensure that a version file with a size `MAX_SNAPSHOT_VERSION_FILE_SIZE + 1` fails

**Comment by brooksprumo** [comment]
Oh, also please re-request a review from me once the changes are made. This way I'll make sure to see the review request in my GH dashboard.

**Comment by mooori** [comment]
Thank you for the review @brooksprumo. The suggestions have been implemented (including tests).

File: runtime/src/snapshot_utils.rs [hunk 3]
```diff
@@ -982,6 +978,28 @@ where
     })
 }
 
+// Reads the `snapshot_version` from a file. Before opening the file, its size is
+// compared to `MAX_SNAPSHOT_VERSION_FILE_SIZE`. If the size exceeds this
+// threshold, it is not opened and an error is returned.
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:983]: Minor: Please make these doc comments (use three slashes: `///`).

File: runtime/src/snapshot_utils.rs [hunk 4]
```diff
@@ -1998,6 +2017,29 @@ mod tests {
...
+        let mut file = NamedTempFile::new().unwrap();
+        file.write_all(file_content.as_bytes()).unwrap();
+        let version_from_file = snapshot_version_from_file(file.path()).unwrap();
+        assert_eq!(version_from_file, file_content);
+    }
+
+    #[test]
+    fn test_snapshot_version_from_file_over_limit() {
+        // Verify `as usize` below is safe.
+        usize::try_from(MAX_SNAPSHOT_VERSION_FILE_SIZE + 1).unwrap();
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:2032]: Capture this as a variable and use it for the size in the `file_content` array below.

**Review by brooksprumo — CHANGES_REQUESTED** [review]
I like it! Two more small things and then I think it'll be all set!

**Review by brooksprumo — APPROVED** [review]
Looks good! Thanks again for your submission, @mooori!

**Comment by brooksprumo** [comment]
@Mergifyio backport v1.8

**Pull request merged**